### PR TITLE
listener: add integration test for in-place update listener filters order

### LINKED ISSUE #20100: Listener update may choose in place update too aggressively
When comparing the old listener and the new listener config,
[all the repeated fields are considered as set.](https://github.com/envoyproxy/envoy/blob/main/source/server/listener_impl.cc#L946)

An unexpected case is, if a listener filter chain is reordered along with some network filter chain add/deletion,
the listener filter chain update is ignored.

I go through the fields in the top level message listener and find these fields. Please correct me if I am wrong

* filter_chains: set semantic
* listener_filter_chains: ordered_list semantic
* socket_options: set semantic
* access_log: set semantic 

listener_filter_chains need to be compared separately.

CC @kyessenov

### PR DESCRIPTION
**Pull request by soulxu**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>

Commit Message: listener: add integration test for in-place update listener filters order
Additional Description:
Test the case of only the listener filters order was updated, then the envoy will do a in-place update for this case.

Risk Level: low
Testing: integration test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Part of Fixes #20100

### CODE REVIEW

**Comment by soulxu** [comment]
cc @lambdai

File: test/integration/listener_lds_integration_test.cc [hunk 1]
```diff
@@ -738,6 +738,52 @@ class ListenerFilterIntegrationTest : public testing::TestWithParam<Network::Add
           stat_prefix: tcp_stats
           cluster: cluster_0
 )EOF") {}
+
+  void createLdsStream() {
+    AssertionResult result =
+        fake_upstreams_[1]->waitForHttpConnection(*dispatcher_, lds_connection_);
+    RELEASE_ASSERT(result, result.message());
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:745]: Now that you are here, can you switch to ASSERT_TRUE(result)? ASSERT_TRUE/EXPECT_TRUE is more friendly in test case.

File: test/integration/listener_lds_integration_test.cc [hunk 2]
```diff
@@ -738,6 +738,52 @@ class ListenerFilterIntegrationTest : public testing::TestWithParam<Network::Add
...
+
+    auto result3 = lds_connection_->waitForNewStream(*dispatcher_, lds_stream_);
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:747]: where is result2 :)

File: test/integration/listener_lds_integration_test.cc [hunk 3]
```diff
@@ -738,6 +738,52 @@ class ListenerFilterIntegrationTest : public testing::TestWithParam<Network::Add
...
+
+  void sendLdsResponse(const std::vector<envoy::config::listener::v3::Listener>& listener_configs,
+                       const std::string& version) {
+    envoy::service::discovery::v3::DiscoveryResponse response;
+    response.set_version_info(version);
+    response.set_type_url(Config::TypeUrl::get().Listener);
+    for (const auto& listener_config : listener_configs) {
+      response.add_resources()->PackFrom(listener_config);
+    }
+    ASSERT(lds_stream_ != nullptr);
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:760]: ASSERT_NE(nullptr, lds_stream_);

File: test/integration/listener_lds_integration_test.cc [hunk 4]
```diff
@@ -738,6 +738,52 @@ class ListenerFilterIntegrationTest : public testing::TestWithParam<Network::Add
...
           cluster: cluster_0
 )EOF") {}
+
+  void createLdsStream() {
+    AssertionResult result =
+        fake_upstreams_[1]->waitForHttpConnection(*dispatcher_, lds_connection_);
+    RELEASE_ASSERT(result, result.message());
+
+    auto result3 = lds_connection_->waitForNewStream(*dispatcher_, lds_stream_);
+    RELEASE_ASSERT(result3, result3.message());
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:748]: and here

File: test/integration/listener_lds_integration_test.cc [hunk 5]
```diff
@@ -902,9 +948,100 @@ TEST_P(ListenerFilterIntegrationTest, MixNoInspectDataFilterAndInspectDataFilter
...
+        close_connection: false
+        )EOF");
+  std::string data = "hello";
+  std::string data_after_drain = data.substr(2, std::string::npos);
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    // Add the static cluster to serve LDS.
+    auto* lds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+    lds_cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+    lds_cluster->set_name("lds_cluster");
+    ConfigHelper::setHttp2(*lds_cluster);
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:979]: TIL!

File: test/integration/listener_lds_integration_test.cc [hunk 6]
```diff
@@ -902,9 +948,100 @@ TEST_P(ListenerFilterIntegrationTest, MixNoInspectDataFilterAndInspectDataFilter
...
+    lds_api_config_source->set_transport_api_version(envoy::config::core::v3::V3);
+    envoy::config::core::v3::GrpcService* grpc_service = lds_api_config_source->add_grpc_services();
+    setGrpcService(*grpc_service, "lds_cluster", fake_upstreams_[1]->localAddress());
+  });
+  on_server_init_function_ = [&]() {
+    createLdsStream();
+    sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "1");
+  };
+  create_grpc_lds_ = true;
+  use_lds_ = false;
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:1002]: I had the impression that use_lds_ and the new create_grpc_lds_ had the same semantic

File: test/integration/listener_lds_integration_test.cc [hunk 7]
```diff
@@ -902,9 +948,100 @@ TEST_P(ListenerFilterIntegrationTest, MixNoInspectDataFilterAndInspectDataFilter
...
+  sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "2");
+  test_server_->waitForCounterGe("listener_manager.listener_create_success", 2);
+  test_server_->waitForCounterEq("listener_manager.listener_in_place_updated", 1);
+
+  IntegrationTcpClientPtr tcp_client2 = makeTcpConnection(lookupPort("test_listener"));
+  ASSERT_TRUE(tcp_client2->write(data));
+  tcp_client2->waitForDisconnect();
+
+  // Then ensure the whole listener works as expect with enough data.
+  std::string data2 = "helloworld";
```
Comment (lambdai) [inline test/integration/listener_lds_integration_test.cc:1033]: nits: rename to long_data and move it below tcp_client3.

The reason is that client2 links to data and client3 links to data2. It's a little bit confusing.

**Review by lambdai — CHANGES_REQUESTED** [review]
Thank you so much!

Generally LGTM modulo some nits

File: test/integration/listener_lds_integration_test.cc (hunk 1 above)
Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:745]: got it

File: test/integration/listener_lds_integration_test.cc (hunk 2 above)
Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:747]: oops, hah, counting mistake :)

File: test/integration/listener_lds_integration_test.cc (hunk 3 above)
Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:760]: got it

File: test/integration/listener_lds_integration_test.cc (hunk 5 above)
Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:979]: sorry, I didn't get what you mean here.

File: test/integration/listener_lds_integration_test.cc (hunk 6 above)
Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:1002]: yes, I can use `use_lds_` only here.

File: test/integration/listener_lds_integration_test.cc (hunk 7 above)
Comment (soulxu) [reply test/integration/listener_lds_integration_test.cc:1033]: yea, let me rename it.

File: test/integration/listener_lds_integration_test.cc (hunk 5 above)
Comment (lambdai) [reply test/integration/listener_lds_integration_test.cc:976]: never mind. I didn't know this helper function. Thanks！

**Review by ggreenway — APPROVED** [review]

**Pull request merged**